### ISSUE TITLE
[Bug] When Using custom annotation, the route will be incorrect

### ISSUE THREAD

**Issue by shuitai** [zandero/rest.vertx#55]
Annotation  
`
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface AuditLog {
    String title() default "";
}
`

Guice Interceptor

`
@Slf4j
public class LogInterceptor implements MethodInterceptor {

    @Override
    public Object invoke(MethodInvocation invocation) throws Throwable {
        TimeInterval timer = DateUtil.timer();
        timer.start();
        Object result = invocation.proceed();
        long costTime = timer.intervalMs();
        log.info("cost time: {}", costTime);
        return result;
    }
}
`
Module
`
public class AdminModule extends AbstractModule {

    @Override
    protected void configure() {
        LogInterceptor logInterceptor = new LogInterceptor();
        requestInjection(logInterceptor);
        bindInterceptor(Matchers.any(), Matchers.annotatedWith(AuditLog.class), logInterceptor);
    }
}
`

API
`
@Slf4j
@Path("/system/user")
public class SysUserApi {
@GET
    @Path("/list")
    @Produces("application/json")
    @RolesAllowed("system:user:list")
    public TableDataInfo list(@QueryParam("user") SysUser user, @Context HttpServerRequest request) {
        startPage(request);
        List<SysUser> list = userService.selectUserList(user);
        return getDataTable(list);
    }

@GET
    @Path("/export")
    @Produces("application/json")
    @RolesAllowed("system:user:export")
    @AuditLog(title = "export")
    public AjaxResult export(@QueryParam("user") SysUser user) {
        List<SysUser> list = userService.selectUserList(user);
        ExcelUtil<SysUser> util = new ExcelUtil<SysUser>(SysUser.class);
        return util.exportExcel(list, "data", downloadPath);
    }
`

Result:
version, rest vertx 0.8.9
1. without the @AuditLog, the route is correct.
[INFO] Registering route:      GET /system/user/export 
[INFO] Registering route:      GET /system/user/list  
2. with the the @AuditLog, the route is incorrect.
[INFO] Registering route:      GET /system/user/export
[INFO]  Registering route:      GET /list/list
3. all route has @AuditLog, the route is correct too.
[INFO] Registering route:      GET /system/user/export 
[INFO] Registering route:      GET /system/user/list

**Comment by drejc** [comment]
Not sure how this should be possible ... rest.vertx does not care for unknown annotations.
Can you provide a working example (simple java project - just push it to GitHub) where this issue is replicated. The "snippets" you have posted don't make much sense to me plus the formatting is messed up.  I assume there is some other issue with your set up.

PS: I have tried to replicate this with the exact same @AuditLog annotation and it works as expected ... but my setup might be different then yours

**Comment by shuitai** [comment]
Please check https://github.com/shuitai/vertx-rest-example/tree/master

**Comment by shuitai** [comment]
https://github.com/shuitai/vertx-rest-example/blob/master/src/main/java/com/vertx/example/api/TestRest.java

**Comment by drejc** [comment]
Very interesting ... not sure what is causing this ... will take a closer look

**Comment by drejc** [comment]
I took a quick look an it seems the problem is caused by Guice altering the injected class annotation signature.

If you disable the following line in GuiceAdminModule:

`// bindInterceptor(Matchers.any(), Matchers.annotatedWith(AuditLog.class), logInterceptor);`

**rest.vertx** works as expected.

Not sure if this can be "fixed" on **rest.vertx** side, as **rest.vertx** relies on an intact class signature in order to collect the annotations. 

Will take a deeper dive and keep you posted.

**Comment by shuitai** [comment]
yes, Remove the bindInterceptor, the issue will not happen, so ituse the custom annotation with guice, it will not work.

**Comment by shuitai** [comment]
Could

**Comment by shuitai** [comment]
In my scenario, we need AOP and define custom annotations to measure the cost time of function, or record some operations in API, I think it is very useful feature. Hope you could find way to fix it. Thanks.

**Comment by shuitai** [comment]
About the bug, I have submitted a pull request, please review https://github.com/zandero/rest.vertx/pull/58

**Comment by drejc** [comment]
Until I push the fix ... here is a code snippet that fixes the issue.

Replace it in AnnotationProcessor.java

```java
/**
     * Gets all route definitions for base class / interfaces and inherited / abstract classes
     *
     * @param clazz to inspect
     * @return collection of all definitions
     */
    private static Map<RouteDefinition, Method> collect(Class<?> clazz) {

        // Addressing issue #55, where Guice alters annotations
        if (clazz.getName().contains("$$EnhancerByGuice$$")) { // Skip Guice enhanced clazz and go to "original"
            clazz = clazz.getSuperclass();
        }

        Map<RouteDefinition, Method> out = getDefinitions(clazz);
        RouteDefinition classDefinition = getClassDefinition(out);

        for (Class<?> inter : clazz.getInterfaces()) {
            Map<RouteDefinition, Method> found = collect(inter);

            RouteDefinition interfaceClassDefinition = getClassDefinition(found);
            if (classDefinition != null) {
                classDefinition.join(interfaceClassDefinition);
            }

            join(out, found);
        }

        Class<?> superClass = clazz.getSuperclass();
        if (superClass != Object.class && superClass != null) {
            Map<RouteDefinition, Method> found = collect(superClass);

            RouteDefinition superClassDefinition = getClassDefinition(found);
            if (classDefinition != null) {
                classDefinition.join(superClassDefinition);
            }

            join(out, found);
        }

        //
        out.put(classDefinition, null);
        return out;
    }
```

**Comment by shuitai** [comment]
@ drejc , thanks, look forward for your fix, when will you release the v0.9?

**Comment by drejc** [comment]
Can't promise anything ... but hopefully in the next week or two
